app/gui2/src/components/ComponentBrowser/filtering.ts

```diff
@@ -129,20 +191,38 @@ class FilteringQualifiedName {
     this.pattern = pattern
     // Starting at some segment, each segment should start with the respective
     // pattern's segment. See `Filtering` docs for full algorithm description.
-    const segmentsMatch = '(^|\\.)' + pattern.replace(/\./g, '[^\\.]*\\.')
+    // For both regexes below:
+    // The first match (`match[1]`) is the part before the first matched input.
+    // The rest of the matches come in pairs:
+    // - The matched segment
+    // - The unmatched part before the next matched segment
+    const segmentsMatch = '(^|.*?[.])(' + pattern.replace(/[.]/g, ')([^.]*)([.]') + ')'
     // The direct members must have no more segments in their path.
-    this.memberRegex = new RegExp(segmentsMatch + '[^\\.]*$', 'i')
-    this.memberOfAnyDescendantRegex = new RegExp(segmentsMatch, 'i')
+    this.memberRegex = new RegExp(segmentsMatch + '([^.]*$)', 'i')
+    this.memberOfAnyDescendantRegex = new RegExp(segmentsMatch + '(.*)', 'i')
   }
 
-  matches(entry: SuggestionEntry, alsoFilteringByPattern: boolean): boolean {
+  private static matchRanges(match: RegExpExecArray) {
+    const result: MatchRange[] = []
+    for (let i = 1, pos = 0; i < match.length; i += 1) {
+      // Matches come in groups of two, and the first matched part is `match[2]` (= 0 mod 2).
+      if (i % 2 === 0) {
+        result.push({ start: pos, end: pos + match[i]!.length })
+      }
+      pos += match[i]!.length
+    }
+    return result
+  }
+
+  matches(entry: SuggestionEntry, alsoFilteringByPattern: boolean): Opt<PartialMatchResult> {
```

Comment:
In return values, I'd prefer to use `PartialMatchResult | null` or `PartialMatchResult | undefined`, instead of Opt. Opt should be rather for arguments: it's convenient when input arguments has as broad types as possible, while output is narrowed as much as possible.

Of course, fix the old methods if needed.

Also, I think that `null` has a bit better semantics here than `undefined`; the latter would suggest that nobody defined the match, but it is well-defined that there is no match.

Reply:
ah. i was assuming we were going with `undefined`, because in `opt.ts` it *does* say:
> In our own code, we should return `undefined` for "no value", since that is the default value for empty or no `return` expression

but that's easy enough to change. the docs should probably be changed eventually, *but* i think it's something that should be discussed with everyone before actually doing it